content/articles/building-autocorrect -feature-using-nlp-with-python/index.md

```diff
@@ -0,0 +1,272 @@
+
+```python
+
+# Now we implement 4 edit word functions
+
+# DeleteLetter:removes a letter from a given word
+def DeleteLetter(word):
+    delete_list = []
+    split_list = []
+    for i in range(len(word)):
+        split_list.append((word[0:i], word[i:]))
+    for a, b in split_list:
+        delete_list.append(a + b[1:])
+    return delete_list
+
+
+delete_word_l = DeleteLetter(word="cans")
+
+# SwitchLetter:swap two adjacent letters
+def SwitchLetter(word):
+    split_l = []
+    switch_l = []
+    for i in range(len(word)):
+        split_l.append((word[0:i], word[i:]))
+    switch_l = [a + b[1] + b[0] + b[2:] for a, b in split_l if len(b) >= 2]
+    return switch_l
+
+
+switch_word_l = SwitchLetter(word="eta")
+
+# replace_letter: changes one letter to another
+def replace_letter(word):
+    split_l = []
+    replace_list = []
+    for i in range(len(word)):
+        split_l.append((word[0:i], word[i:]))
+    alphabets = 'abcdefghijklmnopqrstuvwxyz'
+    replace_list = [a + l + (b[1:] if len(b) > 1 else '') for a, b in split_l if b for l in alphabets]
+    return replace_list
+
+
+replace_l = replace_letter(word='can')
+
+# insert_letter: adds additional characters
+def insert_letter(word):
+    split_l = []
+    insert_list = []
+    for i in range(len(word) + 1):
+        split_l.append((word[0:i], word[i:]))
+    letters = 'abcdefghijklmnopqrstuvwxyz'
+    insert_list = [a + l + b for a, b in split_l for l in letters]
+    # print(split_l)
+    return insert_list
+
+
+
+# combining the edits
+# switch operation optional
+def edit_one_letter(word, allow_switches=True):
+    edit_set1 = set()
+    edit_set1.update(DeleteLetter(word))
+    if allow_switches:
+        edit_set1.update(SwitchLetter(word))
+    edit_set1.update(replace_letter(word))
+    edit_set1.update(insert_letter(word))
+    return edit_set1
+
+
+
+# edit two letters
+def edit_two_letters(word, allow_switches=True):
+    edit_set2 = set()
+    edit_one = edit_one_letter(word, allow_switches=allow_switches)
+    for w in edit_one:
+        if w:
+            edit_two = edit_one_letter(w, allow_switches=allow_switches)
+            edit_set2.update(edit_two)
+    return edit_set2
+
+
+
+# get corrected word
+def get_corrections(word, probs, vocab, n=2):
+    suggested_word = []
+    best_suggestion = []
+    suggested_word = list(
+        (word in vocab and word) or edit_one_letter(word).intersection(vocab) or edit_two_letters(word).intersection(
+            vocab))
+    best_suggestion = [[s, probs[s]] for s in list(reversed(suggested_word))]
+    return best_suggestion
+
+
+my_word = input("Enter any word:")
+probs = get_probs(word_count)
+tmp_corrections = get_corrections(my_word, probs, v, 2)
+for i, word_prob in enumerate(tmp_corrections):
+    print(f"word {i}: {word_prob[0]}, probability {word_prob[1]:.6f}")
```

Comment:
Can we break this big code block down into several code blocks for each method?

And, you could explain each of them below the code snippet.